[2단계 - 로또] 이진호 미션 제출합니다.

binary-ho opened the pull request.

binary-ho committed:
refactor : Numbers를 불필요하게 하위클래스로 선언한 부분 변경

binary-ho committed:
refactor : LottoNumbersDto의 범위에 벗어나는 책임 제거

binary-ho committed:
feat : 로또에 사용되는 숫자의 범위를 한정하는 객체 LottoNumberRange 구현

binary-ho committed:
test : 기존에 범위를 입력받던 객체들이 LottoNumberRange를 통해 숫자가 범위안에 드는지를 확인하는 테스트 코드 구현

binary-ho committed:
refactor : 기존에 범위를 입력받던 객체들이 LottoNumberRange 객체를 입력받아 숫자가 로또 게임 숫자의 범위안에 드는지 확인하도록 변경

binary-ho committed:
test : 구입 갯수를 숫자로 변환하는 테스트 코드 작성

binary-ho committed:
feat : 구입 갯수를 숫자로 변환하는 기능 구현

binary-ho committed:
test : 입력 받은 구입 갯수가 숫자가 아닌 경우 예외를 던지는 테스트 코드 작성

binary-ho committed:
feat : 입력 받은 구입 갯수가 숫자가 아닌 경우 예외를 던지는 기능 구현

binary-ho committed:
test : 입력 받은 구입 갯수가 음수인 경우 예외를 던지는 테스트 코드 작성

binary-ho committed:
feat : 입력 받은 구입 갯수가 음수인 경우 예외를 던지는 기능 구현

binary-ho committed:
test : 당첨 번호를 입력 받아 LottoNumbers로 반환하는 테스트 코드 작성

binary-ho committed:
refactor : WinningNumberConvertor의 책임을 로또에서 쓰이는 로또 번호들을 입력 받아 변환하는 책임으로 바꿔주기 위해 이름을 LottoNumberConvertor로 변경

binary-ho committed:
refactor : Numbers를 삭제하고 WinningNumbers를 LottoNumbers가 상속하도록 변경

binary-ho committed:
refactor : 불필요한 ManualLottoNumbersConvertor.java 삭제

binary-ho committed:
feat : InputView에서 수동 로또 구매 갯수와 수동 로또 번호의 입력을 받는 메서드 구현

binary-ho committed:
feat : 로또 구매 갯수를 저장하는 객체 LottoPurchaseCount 구현

binary-ho committed:
feat : 수동 로또 구매 갯수가 구매 로또 갯수를 초과하는 경우 예외를 던지는 기능 구현

binary-ho committed:
feat : 수동 로또 구매 갯수가 구매 로또 갯수를 초과하는 경우 예외를 던지고 다시 입력 받는 기능 구현

binary-ho committed:
feat : OutputView의 로또 구매 갯수 출력문 변경

binary-ho committed:
test : 현재 남은 로또 구매 가능 갯수보다 더 많은 로또 구매시 예외를 던지는 테스트 코드 작성

binary-ho committed:
feat : 로또 티켓을 만들 때, 로또 티켓을 입력 받아 합칠 수 있는 기능 구현

binary-ho committed:
feat : 갯수를 입력 받아, 갯수만큼 수동으로 로또 번호를 입력하여 티켓을 구매할 수 있는 기능 구현

skullkim commented on src/main/java/domain/LottoPurchaseCount.java:
```diff
@@ -0,0 +1,23 @@
+
+    public LottoPurchaseCount decreaseCount(int count) {
+        int result = lottoPurchaseCount - count;
+        if (result < 0) {
```
매직넘버 사용은 자제해 주세요~

binary-ho replied:
확인했습니다! 
상수 MINIMUM_STOCK을 선언하여 이용했습니다.

skullkim commented on src/main/java/controller/LottoController.java:
```diff
@@ -25,22 +31,29 @@ public class LottoController {
+	private int manualLottoPurchaseCount;
+
+	public LottoController() {
+		lottoNumberRange = new LottoNumberRange(FROM, TO);
```
컨트롤러는 어떤 역할을 할까요? 도메인 이라는 것은 무엇일까요? 로또 게임이라는 도메인에서 로또 숫자의 범위가 어디에 위치하는 것이 적절할까요?

binary-ho replied:
컨트롤러는 표현 영역에 속한 객체로 사용자 입력을 받아
응용 영역이나 도메인 영역에 전달하고, 처리 결과를 다시 사용자에게 전달하는 역할을 합니다.

그리고 도메인은 소프트웨어에서 핵심 문제 영역입니다. 가령 로또게임에서는 로또입니다. 
이런 도메인들의 핵심 로직을 구현한 것이 도메인 모델이고, 앞서 언급한 도메인 영역에는 이 도메인 모델들이 위치해 있습니다.

객체지향 패러다임 소프트웨어는 객체들간의 상호 협력에 집중합니다.
오늘 돌아가는 것이 가장 중요하지만, 내일 변경 가능한 것도 중요하므로,
객체지향에서 객체는 응집도가 높아야 하고 서로 의존도가 낮아야 합니다.

저는 책임을 수행할 객체가 없을 때 객체를 만들고,
한 객체가 많은 책임을 갖고 있는 경우, 응집도를 높히기 위해 객체를 나눕니다.
 
물론 무조건 많이 나눈다고 무조건 좋은 것은 아닙니다, 
LottoRange객체를 만들게 된건, 애초에 로또에 쓰이는 다양한 객체들이 범위를 주입받도록 설계하며 시작된 것 같습니다.

주입받도록 만든 이유는 변경을 용이하게 하고 싶었고, 난수 테스트를 쉽게 하고 싶었습니다.
물론 그런 명세도 없고, 실제로 국내에 이런 로또는 없지만, 로또 객체들이 다양한 숫자 범위를 가질 수도 있다고 생각했습니다. 로또는 1 ~ 45의 숫자를 범위로 가지고 있습니다. 하지만 자동 발급 기계는 그럴 필요가 없다고 생각했습니다. 이번 경우엔 자동 발급 기계도 1 ~ 45의 숫자를 발급하고, 보너스볼 허용 범위도 1 ~ 45를 가지지만, 
예를 들어 발급 기계가 10 ~ 30의 숫자만을 발급하거나, 보너스볼은 1 ~ 10 까지만 허용되는 상황이 있을 수도 있다고 생각했습니다.  

그리고 이런 방식이면 난수가 포함된 로또의 테스트 또한 쉽게 할 수 있겠다고 생각했습니다. 
그래서 테스트 코드 또한 1, 2, 3등에 대한 테스트는 1 ~ 6까지만 나오게 해서 1등이 나오게 하고, 1을 보너스 번호로, 2 ~ 7을  뽑아 2등이 나오게 하는 방식으로 구성했었습니다.

하지만 이런 구현은 잘못된 범위가 주입될 가능성이 있다고 말씀주셔서,
해결해보기 위해 Lotto 객체가 직접 불변 범위를 객체를 만들어 주는 방식을 택했습니다.
LottoRange 객체는 이런 배경에서 만들게 되었습니다.

하지만 지금 생각해보면 이는 불완전합니다. LottoRange를 불변으로 선언하여 문제를 해결하겠다고 했지만, 
꼭 LottoRange의 생성이 Lotto객체를 통해서만 이루어질 수 있는 것이 아니기 때문에, 
직접 생성자를 호출하여 만들어질 수도 있기 때문에, 얼마든지 잘못된 범위가 들어갈 수 있다는 점은 바뀌지 않았습니다.

그리고 범위를 주입받는 설계는 오버엔지니어링일 수 있습니다. 
그런 명세는 아직 없었고, 다른 구현도 충분히 가능했기 때문입니다.

그래서 지금은 LottoNumber 객체가 상수로 로또 숫자의 범위나 갯수 제한을 알고 있고, 
당첨 번호나 보너스 볼은 LottoNumber 객체를 컴포지션으로 가지게 함으로서 이 제한을 공유하고,
LottoGenerator는 LottoNumber의 상수를 활용하는 방식으로 구현했습니다. 
(원래는 LottoNumber 객체가 LottoGenerator객체에 직접 범위를 주입해주면서 생성하도록 구현했다가 앞서 언급한 것과 같이 외부에서 생성자 호출로 만들수 있다는 점을 인지하여 이 글을 작성하며 수정했습니다.)

skullkim commented on src/main/java/domain/numbers/WinningNumbers.java:
```diff
@@ -2,17 +2,15 @@
-public class WinningNumbers extends Numbers {
-
-	private static final int WINNING_NUMBER_SIZE = 6;
+public class WinningNumbers extends LottoNumbers {
```
상속은 어떤 이유로 존재하는 걸까요? LottoNumbers가 추상 클래스가 아닌 일반 클래스네요. 일반 클래스를 extends하면 어떤 문제가 존재할까요?

skullkim commented on src/main/java/domain/numbers/LottoNumbers.java:
```diff
@@ -2,32 +2,42 @@
 import java.util.List;
 
-public class LottoNumbers extends Numbers {
+public class LottoNumbers {
```
상속과 조합은 각각 어떤 상황에서 사용해야 할까요? [상속보다는 조합(Composition)을 사용하자](https://tecoble.techcourse.co.kr/post/2020-05-18-inheritance-vs-composition/)를 참고해보세요~

binary-ho replied:
참고했습니다. 덕분에 컴포지션에 대해 공부하고 적용해볼 기회를 가졌습니다. 감사합니다!
어째서인지 딱 저 아이템만 빼고 해당 챕터를 읽었네요 

상속을 통해 구현된 하위클래스는 
1. 캡슐화를 깨트리고,
2. 상위 클래스의 자기 사용여부 파악이 힘들어 제대로 구현하기도 힘들며,
3. 변화에 취약합니다.

상위 클래스의 메서드들이 내부적으로 자신의 메서드를 사용하며 구현된 것을 자기 사용이라고 합니다.
상위 클래스가 자기 사용을 통해 구현했다면, 재정의하기가 매우 까다로워집니다.
이런 자기 사용 메서드들의 쓰임을 모두 이해해야 제대로된 재정의가 가능하고, 오동작하는 메서드를 만들기가 쉽습니다.

그리고 변경과 추가에 약합니다. 
확장 목적으로 클래스를 상속해 구현한다면, 상위 클래스의 내부 구현이 변경될 경우 쉽게 흔들릴 수 있게 됩니다.
최악의 경우 메서드가 추가되면서 컬렉션 원소들의 검증 과정에 구멍이 생길 수도 있습니다.

컴포지션은 상속할 클래스를 내부 필드로 사용하는 방식으로,
이를 통해 위의 문제들을 피해갈 수 있습니다.

skullkim commented on src/main/java/domain/LottoNumberRange.java:
```diff
@@ -0,0 +1,24 @@
+package domain;
+
+public class LottoNumberRange {
```
객체를 만드는 이유가 뭘까요?

skullkim reviewed (changes requested):
멘토링 마지막 단계네요~. 끝까지 잘 마무리 지어봐요!

binary-ho committed:
feat : 매직넘버의 상수화

binary-ho committed:
refactor : WinningNumbers가 LottoNumber를 상속하지 않고 컴포지션으로 사용.
> BonusBall은 WinningNubers의 int 필드로 사용 BonusBall이 갖던 책임을 WinningNumbers에 위임
> 불필요해진 BonusBall 객체와 LottoNumberRange 객체 삭제

binary-ho commented:
3번의 커밋으로 끊어야 할 일을 한번에 해버렸습니다.
읽기 불편하실 것 같습니다. 죄송합니다.

binary-ho committed:
refactor : LottoGenerator가 로또 숫자의 범위나 갯수 제한을 주입받지 않고, LottoNumber 객체의 상수를 사용하는 방식으로 변경
